### PR TITLE
Adds support for automatic retry on 502, 503, and 500 errors + refactoring of duplicated code

### PR DESCRIPTION
**Pull request by petrroll**
Hi, I've had some problems with Spotify API returning 502 errors randomly on all types of requests so I thought that adding an automatic retry support into the library might be a good idea. 

It's off by default and it's fully configurable. You can specify the maximum number of tries and the time between individual attempts. 

It's implemented for all requests both in synchronous and async variants.

### CODE REVIEW

**Comment by petrroll** [comment]
So far the configuration support has not yet been implemented into WebAPI factory as I'm not sure how to do it properly (-> in a way it doesn't clash with your design ideas).

File: SpotifyAPI/Web/SpotifyWebAPI.cs [hunk 1]
```diff
@@ -44,6 +44,25 @@ public void Dispose()
...
         }
 
+        #region Configuration
+        /// <summary>
+        /// Specifies after how many miliseconds should a failed request be retried.
+        /// </summary>
+        public int RetryAfter { get; set; } = 50;
+
+        /// <summary>
+        /// Should a failed request (Error 500, 502, or 503) be automatically retried or not.
```
Comment (JohnnyCrazy) [inline SpotifyAPI/Web/SpotifyWebAPI.cs]: A reference to the `RetryErrorCodes`-property would be better than `500, 502 etc.`

File: SpotifyAPI/Web/SpotifyWebAPI.cs [hunk 2]
```diff
@@ -44,6 +44,25 @@ public void Dispose()
...
+        /// </summary>
+        public bool UseAutoRetry { get; set; } = false;
+
+        /// <summary>
+        /// Maximum number of tries for one failed request.
+        /// </summary>
+        public int RetryTimes { get; set; } = 10;
+
+        public IEnumerable<int> RetryErrorCodes { get; private set; } = new int[] { 500, 502, 503 };
```
Comment (JohnnyCrazy) [inline SpotifyAPI/Web/SpotifyWebAPI.cs]: Docs would be a +1 :)

File: SpotifyAPI/Web/SpotifyWebAPI.cs [hunk 3]
```diff
@@ -1911,51 +1826,161 @@ public Paging<T> GetPreviousPage<T>(Paging<T> paging)
             return await GetPreviousPageAsync<Paging<T>, T>(paging);
         }
 
+        private ListResponse<bool> DownloadDataAltAndExtract(string url)
```
Comment (JohnnyCrazy) [inline SpotifyAPI/Web/SpotifyWebAPI.cs]: How about making this little more generic, I would suggest:
`private ListResponse<T> DownloadList<T>(string url)`

Same for other methods

**Comment by JohnnyCrazy** [comment]
Hi,

Thanks for the PR, good idea!
I'm personaly not a fan of Auto-Retry, especially for 5XX error codes, since these shouldn't happen and spotify should take care of it. (Did you tell them btw.?) But I guess nothing is wrong with implementing it while it's disabled by default.

I left some comments, maybe you can take a look @petrroll

File: SpotifyAPI/Web/SpotifyWebAPI.cs (hunk 1 above)
Comment (petrroll) [inline SpotifyAPI/Web/SpotifyWebAPI.cs]: Done.

File: SpotifyAPI/Web/SpotifyWebAPI.cs (hunk 2 above)
Comment (petrroll) [inline SpotifyAPI/Web/SpotifyWebAPI.cs]: I'm terrible at writing docs but I'll try to look at it.

File: SpotifyAPI/Web/SpotifyWebAPI.cs (hunk 3 above)
Comment (petrroll) [inline SpotifyAPI/Web/SpotifyWebAPI.cs]: Ok, I'll take a look at it. Probably won't have time before Thursday though.

File: SpotifyAPI/Web/SpotifyWebAPI.cs (hunk 2 above)
Comment (JohnnyCrazy) [inline SpotifyAPI/Web/SpotifyWebAPI.cs]: Just a inline doc @petrroll , like the other properties

Comment (petrroll) [inline SpotifyAPI/Web/SpotifyWebAPI.cs]: @JohnnyCrazy  Hmm, I don't think I follow now. The only public properties/methods I've added are the configuration ones that are already "inline" documented. 

I've made no other changes to the public API of the library. I only changed some Utils method but non of them was documented before...

Or do you mean that I should just copy the "inline" docs to the proper documentation?

Comment (JohnnyCrazy) [inline SpotifyAPI/Web/SpotifyWebAPI.cs]: No, I'm just saying the property `public IEnumerable<int> RetryErrorCodes` is missing XML Documentation

File: SpotifyAPI/Web/SpotifyWebAPI.cs (hunk 3 above)
Comment (petrroll) [inline SpotifyAPI/Web/SpotifyWebAPI.cs]: @JohnnyCrazy It was actually easier than I expected, done.

File: SpotifyAPI/Web/SpotifyWebAPI.cs (hunk 2 above)
Comment (petrroll) [inline SpotifyAPI/Web/SpotifyWebAPI.cs]: Ou, I see now :). Fixed in 2 minutes.

File: SpotifyAPI/Web/SpotifyWebAPI.cs [hunk 4]
```diff
@@ -44,6 +44,28 @@ public void Dispose()
...
+
+        /// <summary>
+        /// Maximum number of tries for one failed request.
+        /// </summary>
+        public int RetryTimes { get; set; } = 10;
+
+        /// <summary>
+        /// Error codes that will trigger auto-retry if <see cref="UseAutoRetry"/> is enabled.
+        /// </summary>
+        public IEnumerable<int> RetryErrorCodes { get; private set; } = new int[] { 500, 502, 503 };
```
Comment (JohnnyCrazy) [inline SpotifyAPI/Web/SpotifyWebAPI.cs]: Any reason for the `private set`?

Comment (petrroll) [inline SpotifyAPI/Web/SpotifyWebAPI.cs]: Hmm, looking at it... not really. For some reason I thought it shouldn't be a part of public API but can't remember why right now.

**Comment by JohnnyCrazy** [comment]
Thanks again! :)

**Pull request merged**